lib/chef/provider/git.rb

```diff
@@ -50,6 +50,12 @@ def define_resource_requirements
             "Cannot clone #{new_resource} to #{cwd}, the enclosing directory #{dirname} does not exist")
         end
 
+        requirements.assert(:checkout) do |a|
+          a.assertion { target_dir_non_existent_or_empty? }
```

Comment:
@lamont-granquist Do we want to take why-run into account here?

Reply:
Good question. I’ve heard rumors that why run is inherently flawed and possibly going away anyway in the future? I didn’t really think about it TBH.

Reply:
Using "why-run" to simulate chef runs is inherently flawed, yes. However, the chef client uses why-run internally to make decisions about when to run `:before` notifications, among other things.

Reply:
Looping back on this today. What's the plan here? Should we raise an assertion during why run or not?

Reply:
@brownmike why-run is used in :before notifications and now can't really ever go away

Reply:
> why-run is used in :before notifications and now can't really ever go away

I'm not sure what this means in the context of this review, though. Should I add an assertion during why run? Or just a message like some of the other requirements. i.e. the parent dir must exist requirement just has this:

```ruby
a.whyrun("Directory #{dirname} does not exist, this run will fail unless it has been previously created. Assuming it would have been created.")
```

Reply:
Yeah, that's more or less what you need although i think if its non-existent or empty that's fine, so the message needs to be along the lines of "Directory is not a git repository, assuming someone will have cleaned that up" or "Directory contains some kind of garbage, assuming someone cleaned that up..."

The way the whyrun message is pretty oddly worded suggests that it may be better to have the git resource own the target directory and blow it away and assert control over the entire state (which is properly declarative).